python/hail/expr/functions.py

```diff
@@ -1896,7 +1896,20 @@ def is_transversion(ref, alt) -> BooleanExpression:
     -------
     :class:`.BooleanExpression`
     """
-    return _func("is_transversion", tbool, ref, alt)
+    return is_snp(ref, alt) & (~(_is_snp_transition(ref, alt)))
+
+
+@typecheck(ref=expr_str, alt=expr_str)
+def _is_snp_transition(ref, alt) -> BooleanExpression:
+    i = 0
+    while i < hl.eval_expr(ref.length()):
+        if hl.eval_expr(ref[i] == alt[i]):
+            i += 1
+        else:
+            return (((ref[i] == 'A') & (alt[i] == 'G')) |
+                    ((ref[i] == 'G') & (alt[i] == 'A')) |
+                    ((ref[i] == 'C') & (alt[i] == 'T')) |
+                    ((ref[i] == 'T') & (alt[i] == 'C')))
```

Comment:
You may want a default return value, in case someone decides to test `hl.is_transition(hl.literal('A'), hl.literal('A'))`

Reply:
Actually this is kinda funky. Current behavior seems a little odd:
```
hl.eval_expr(hl.is_snp(hl.literal('A'), hl.literal('A')))
True
```
But
```
hl.eval_expr(hl.is_transition(hl.literal('A'), hl.literal('A')))
```
Throws:
```
Error summary: StringIndexOutOfBoundsException: String index out of range: 1
```
I think this PR will fix the error, but just want to make sure you're doing the right thing with poorly behaved sites.

Reply:
I think probably `['A', 'A']` just shouldn't be considered a snp, right? in which case the current logic is probably sufficient, but we'll need to fix `allele_type` to account for that.

Reply:
I agree (well, the current logic is wrong then since it returns `True`, but otherwise yes)

Reply:
I changed `hl.allele_type('A', 'A')` to return `Unknown` instead of `SNP`.